src/version_set.rs

```diff
@@ -1,59 +1,67 @@
 // SPDX-License-Identifier: MPL-2.0
 
-//! As its name suggests, the [VersionSet] trait describes sets of versions.
-//!
-//! One needs to define
-//! - the associate type for versions,
-//! - two constructors for the empty set and a singleton set,
-//! - the complement and intersection set operations,
-//! - and a function to evaluate membership of versions.
-//!
-//! Two functions are automatically derived, thanks to the mathematical properties of sets.
-//! You can overwrite those implementations, but we highly recommend that you don't,
-//! except if you are confident in a correct implementation that brings much performance gains.
-//!
-//! It is also extremely important that the `Eq` trait is correctly implemented.
-//! In particular, you can only use `#[derive(Eq, PartialEq)]` if `Eq` is strictly equivalent to the
-//! structural equality, i.e. if version sets have canonical representations.
-//! Such problems may arise if your implementations of `complement()` and `intersection()` do not
-//! return canonical representations so be careful there.
-
 use std::fmt::{Debug, Display};
 
 use crate::Ranges;
 
-/// Trait describing sets of versions.
+/// A set of versions.
+///
+/// See [`Ranges`] for an implementation.
+///
+/// The methods with default implementations can be overwritten for better performance, but their
+/// output must be equal to the default implementation.
+///
+/// # Equality
+///
+/// It is important that the `Eq` trait is implemented so that if two sets contain the same
+/// versions, they are equal under `Eq`. In particular, you can only use `#[derive(Eq, PartialEq)]`
+/// if `Eq` is strictly equivalent to the structural equality, i.e. if version sets are always
+/// stored in a canonical representations. Such problems may arise if your implementations of
+/// `complement()` and `intersection()` do not return canonical representations.
+///
+/// For example, `>=1,<4 || >=2,<5` and `>=1,<4 || >=3,<5` are equal, because they can both be
```

Comment:
Perhaps explicitly call out that this  example is for integers.

Reply:
I thought about them in the context of PEP 440 and semver, are there cases when this doesn't hold?